pyrefly/lib/state/lsp.rs

```diff
@@ -2861,24 +2896,100 @@ impl<'a> Transaction<'a> {
                 // todo(kylei): optimization, avoid duplicate ast walks
                 if let Some(mod_module) = self.get_ast(handle) {
                     let nodes = Ast::locate_node(&mod_module, position);
+
+                    // If no node covers the current position, try to find a preceding import statement
+                    let mut processed_as_import_statement = false;
                     if nodes.is_empty() {
-                        self.add_keyword_completions(handle, &mut result);
-                        self.add_local_variable_completions(handle, None, position, &mut result);
-                        self.add_builtins_autoimport_completions(handle, None, &mut result);
+                        for stmt in mod_module.body.iter().rev() {
```

Comment:
The logic in lsp.rs is very difficult for me to follow. I would prefer if we could avoid doing string slicing to figure out whether we're in an import or not.

Based on poking around with the ruff AST, https://play.ruff.rs/c109ef63-0112-4318-b7d0-031fce65d738 it seems like trailing space imports generate a StmtImport or StmtImportFrom with `names=[]`. Would it be possible to check that instead?

Reply:
Also, it's a bit odd we iterate thru the statements in reverse when it's more likely to find import statements at the top

Reply:
I would prefer the AST logic to be extracted into a helper like `locate_immediately_preceding_import` in the same module as `Ast::locate_node`